railib/api.py

```diff
@@ -171,46 +203,68 @@ def _get_collection(ctx, path: str, key=None, **kwargs):
     rsp = json.loads(rsp.read())
     return rsp[key] if key else rsp
 
-
-# Parses "multipart/form-data" responses. It returns the parts in a list.
-def _parse_multipart(content_type: str, content: bytes) -> list:
+# Parse "multipart/form-data" response
+def _parse_multipart_form(content_type: str, content: bytes) -> List[TransactionAsyncFile]:
     result = []
-    content_type_json = b'application/json'
-    content_type_arrow = b'application/vnd.apache.arrow.stream'
-    boundary = _extract_multipart_boundary(content_type)
-
-    parts = content.split(b'\r\n' + boundary)
-    for i, part in enumerate(parts):
-        # fix the first part, it may contain the boundary
-        if i == 0:
-            part = part.split(boundary)[-1]
-        # Skip "--\r\n"
-        if b'--\r\n' in part:
-            continue
-        # Part body and headers are separated with CRLFCRLF. Get the body.
-        part_value = part.split(b'\r\n\r\n')[-1]
-        if content_type_json in part:
-            result.append(json.loads(part_value))
-        # if the part has arrow stream, then decode the arrow stream
-        # the results are in a form of a tuple/table
-        elif content_type_arrow in part:
-            with pa.ipc.open_stream(part_value) as reader:
-                schema = reader.schema
-                batches = [batch for batch in reader]
-                table = pa.Table.from_batches(batches=batches, schema=schema)
-                result.append(table.to_pydict())
 
-    return result
+    parts = multipart.decoder.MultipartDecoder(
+        content_type=content_type, content=content).parts
 
+    for part in parts:
+        txn = TransactionAsyncFile()
+        txn.content_type = part.headers[b'Content-Type'].decode()
+        txn.content = part.content
 
-# Finds and returns the multipart form-data boundary.
-# Example:
-#   "Content-Type: multipart/form-data; boundary=d6d47cadd395db7f6648a2ffb65751eb"
-def _extract_multipart_boundary(content_type: str) -> bytes:
-    if "boundary=" not in content_type:
-        raise Exception("no multipart boundary")
-    return b'--' + content_type.split("boundary=")[-1].encode("utf-8")
+        disposition = part.headers[b'Content-Disposition']
+        name = re.match(b'.*; name="(.+?)"', disposition)
+        if not(name is None):
+            txn.name = name.group(1).decode()
+        filename = re.match(b'.*filename="(.+?)"', disposition)
+        if not(filename is None):
+            txn.filename = name.group(1).decode()
```

Comment:
bummer that requests-toolbelt doesn't parse this for us, but this seems like a good workaround…

Maybe we could call this variable `txn_file` or something instead of `txn`?